src/mainboard/sunxi/nezha/main/src/main.rs

```diff
@@ -21,53 +21,47 @@ use spin;
 
 mod sbi_platform;
 
-const MEM: usize = 0x4000_0000;
+const RAM_BASE: usize = 0x4000_0000;
 
 // see ../fixed-dtfs.dts
-// const PAYLOAD_OFFSET: usize = 0x2_0000;
-const PAYLOAD_SIZE: usize = 0x20_0000; // 2 MB
-const PAYLOAD_ADDR: usize = MEM + 0x20_0000;
+// target location for decompressed image
+const PAYLOAD_OFFSET: usize = 0x20_0000;
+const PAYLOAD_ADDR: usize = RAM_BASE + PAYLOAD_OFFSET;
+const PAYLOAD_SIZE: usize = 0x0200_0000; // 32 MB
 
-// compressed image
-const LINUXBOOT_TMP_OFFSET: usize = 0x0400_0000;
-const LINUXBOOT_TMP_ADDR: usize = MEM + LINUXBOOT_TMP_OFFSET;
+const DTB_ADDR: usize = PAYLOAD_ADDR + PAYLOAD_SIZE;
+const DTB_SIZE: usize = 0x0001_0000;
 
-// target location for decompressed image
-const LINUXBOOT_OFFSET: usize = 0x0020_0000;
-const LINUXBOOT_ADDR: usize = MEM + LINUXBOOT_OFFSET;
-const LINUXBOOT_SIZE: usize = 0x0200_0000;
-// DTB_OFFSET should be >=LINUXBOOT_OFFSET+LINUXBOOT_SIZE and match bt0
-// TODO: Should we just copy it to a higher address before decompressing Linux?
-const DTB_OFFSET: usize = LINUXBOOT_OFFSET + LINUXBOOT_SIZE;
-const DTB_ADDR: usize = MEM + DTB_OFFSET;
+const COMPRESSED_ADDR: usize = RAM_BASE + 0x0400_0000;
+const COMPRESSED_SIZE: usize = 0x00fe_0000;
 
 fn decompress_lb() {
     // check for Device Tree header, d00dfeed
     let dtb = unsafe { read_volatile(DTB_ADDR as *const u32) };
     if dtb != 0xedfe0dd0 {
-        panic!("DTB looks wrong: {:08x}\n", dtb);
+        panic!("DTB looks wrong: {dtb:08x}");
     } else {
-        print!("DTB looks fine, yay!\n");
+        println!("DTB looks fine, yay!");
     }
 
     unsafe {
-        decompress(LINUXBOOT_TMP_ADDR, LINUXBOOT_ADDR, LINUXBOOT_SIZE);
+        decompress(COMPRESSED_ADDR, PAYLOAD_ADDR, COMPRESSED_SIZE, PAYLOAD_SIZE);
     }
 
     // check for kernel to be okay
-    let a = LINUXBOOT_ADDR + 0x30;
+    let a = PAYLOAD_ADDR + 0x30;
     let r = unsafe { read_volatile(a as *mut u32) };
     if r == u32::from_le_bytes(*b"RISC") {
-        print!("Payload looks like Linux Image, yay!\n");
+        println!("Payload looks like Linux Image, yay!");
     } else {
-        panic!("Payload does not look like Linux Image: {:x}\n", r);
+        panic!("Payload does not look like Linux Image: {r:x}");
     }
     // Recheck on DTB, kernel should not run into it
```

Comment:
What does this mean? It seems superfluous; if the concern is that the DTB might overlap the payload that you're decompressing above, then a) that's a violation of Rust's memory safety rules, since you're constructing a mutable slice in `decompress` right now that would alias an existing structure, and b) I would test for that specifically, and not as a side-effect of decompression.  Here, you should be able to leverage the language and its guarantees to avoid duplicating the check.

Reply:
Yea, this is no longer necessary; this is from earlier iterations and can be removed now.